src/vcon/MarketGovernance.sol

```diff
@@ -0,0 +1,653 @@
+/// Anytime the rewards share price changes, so does the unclaimed user VCON rewards.
+///
+/// @dev users can stake on PCV deposits that are not productive, such as the holding deposit,
+/// however, they will receive no rewards for doing so if the profit to VCON ratio is not set.
+/// The PSM will not be able to be staked on as it is not whitelisted in the PCV Oracle.
+contract MarketGovernance is CoreRefV2, IMarketGovernance {
+    using DeviationWeiGranularity for *;
+    using SafeERC20 for *;
+    using SafeCast for *;
+
+    /// @notice reference to the PCV Router
+    address public pcvRouter;
+
+    /// @dev convention for all normal mappings is key (venue -> value)
+
+    /// @notice amount of VCON paid per unit of revenue generated per venue
+    /// different venues may have different ratios to account for rewards
+    /// which will not be included in the V1
+    mapping(address => uint256) public profitToVconRatio;
+
+    /// and do the conversion to VCON at the end when accruing rewards
+    /// pack venueLastRecordedProfit, venueTotalShares and profitToVconRatio
+    /// into a single slot for gas optimization
+
+    /// @notice last recorded profit index per venue
+    mapping(address => uint128) public venueLastRecordedProfit;
+
+    /// @notice last recorded VCON share price index per venue
+    mapping(address => uint128) public venueLastRecordedVconSharePrice;
+
+    /// @notice total vcon deposited per venue
+    mapping(address => uint256) public venueTotalShares;
+
+    /// @notice map an underlying token to the corresponding holding deposit
+    mapping(address => address) public underlyingTokenToHoldingDeposit;
+
+    /// no balance checks when unstaking
+
+    /// ---------- Per Venue User Profit Tracking ----------
+
+    /// @dev convention for all double nested address mappings is key (venue -> user) -> value
+
+    /// @notice record how much VCON a user deposited in a given venue
+    mapping(address => mapping(address => uint256)) public venueUserShares;
+
+    /// @param _core reference to core
+    /// @param _pcvRouter reference to pcvRouter
+    constructor(address _core, address _pcvRouter) CoreRefV2(_core) {
+        pcvRouter = _pcvRouter;
+    }
+
+    /// @notice update the VCON share price and the last recorded profit for a given venue
+    /// @param venue address to accrue
+    function accrueVcon(address venue) external globalLock(1) whenNotPaused {
+        IPCVOracle oracle = pcvOracle();
+
+        require(oracle.isVenue(venue), "MarketGovernance: invalid destination");
+
+        _accrue(venue);
+    }
+
+    /// ---------- Permissionless User PCV Allocation Methods ----------
+
+    /// @notice a user can get slashed up to their full VCON stake for entering
+    /// a venue that takes a loss.
+    /// @param amountVcon to stake on destination
+    /// @param destination address to accrue rewards to, and send funds to
+    function stake(
+        uint256 amountVcon,
+        address destination
```

Comment:
interface says `source` but implementation says `destination` for this parameter, it's a bit confusing. There is no PCV movement so you can just call it `venue` imo